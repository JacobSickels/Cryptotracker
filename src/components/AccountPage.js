import React from 'react';
import { connect } from 'react-redux';
import { Input } from 'react-materialize';

import { startEditCurrency } from '../actions/filters';

export class AccountPage extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            currency: props.currency
        }

    }

    onInputChange = (e) => {
        const currency = this.props.exchanges.find((element) => element.id === e.target.value);
        this.setState({ currency });
        this.props.startEditCurrency(currency);
    }

    render() {
        return (
            <div className="page-container">
                <div className="container">
                    <h2>My Account Page</h2>

                    <h1>Set Base Currency</h1>
                    <Input s={5} type='select' label="Select Currency" value={this.state.currency.id} onChange={this.onInputChange}>
                        {
                            this.props.exchanges.map((currency) =>
                                <option key={currency.id} value={currency.id}>{currency.name}</option>
                            )
                        }
                    </Input>
                </div>
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditCurrency: (currency) => dispatch(startEditCurrency(currency))
    };
};

const mapStateToProps = (state, props) => {
    return {
        exchanges: state.exchange,
        currency: state.filters.currency
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(AccountPage);