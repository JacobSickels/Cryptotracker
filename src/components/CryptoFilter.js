import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';

import { setStartDate, setEndDate } from '../actions/filters';

export class CryptoFilter extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    render() {
        return (
            <div className="date-container">
                <div className="container">
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        minimumNights={0}
                        isOutsideRange={(day) => {
                            const today = moment();
                            return day >= today;
                        }}
                    />
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(CryptoFilter);