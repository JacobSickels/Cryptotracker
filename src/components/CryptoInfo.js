import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Row, Col } from 'react-materialize';

import { selectCryptos, getMaxMinData } from '../selectors/cryptos';

export class CryptoInfo extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            maxDifference: () => {
                const crypto = props.data;
    
                let maxDifference = 0;
                let minElement = crypto[0].amount;
                
                for(var i = 1; i < crypto.length; i++) {
                    if(crypto[i].amount - minElement > maxDifference) {
                        maxDifference = crypto[i].amount - minElement;
                    }
                    if(crypto[i] < minElement) {
                        minElement = crypto[i].amount;
                    }
                }
                return maxDifference.toFixed(2);
            }
        }
    }

    render() {
        return (
            <div className="chart-info">
                <div className="container">
                    <Row> 
                        <Col s={12} m={4} l={4}>
                            <h1> Max Increase </h1>
                            {this.props.maxDifference}
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Buy</h1> 
                            <em>{moment(this.props.minElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.minElement.amount}</p>
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Sell</h1>
                            <em>{moment(this.props.maxElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.maxElement.amount}</p>
                        </Col>
                    </Row>
                </div>
            </div>
        );
    }

}

const mapStateToProps = (state, props) => {
    const data = getMaxMinData(state.cryptos, state.filters, props.name.toLowerCase());
    return { ...data }
};

export default connect(mapStateToProps)(CryptoInfo);