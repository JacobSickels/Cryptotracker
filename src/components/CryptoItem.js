import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import { connect } from 'react-redux';
import { ResponsiveContainer, LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';
import selectCryptos from '../selectors/cryptos';

export class CryptoItem extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            showTrendLine: true
        }
    }

    render() {
        return (
            <div className="chart">
                <div className="chart-title">
                    <div className="container">
                        <div className="chart-title--container">
                            <div>
                                <Link className="chart-title__link" to={`/filter/${this.props.name.toLowerCase()}`} >
                                    <h1> {this.props.name} </h1>
                                </Link>
                            </div>
                            <div className="chart-title__amount">
                                <h1>
                                    {this.props.currency_symbol} {this.props.current.amount}
                                </h1>
                            </div>
                            <div className="chart-title--percentage">
                                 <h1 style={{
                                     color:
                                        (this.props.percentage >= 0) ? 'green' : 'red'
                                     }}>
                                    {this.props.percentage.toFixed(2)}% 
                                </h1>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="container">
                    <div className="button-container">
                        <button className="button--trend"
                            onClick={() => {
                                const show = this.state.showTrendLine;
                                this.setState({showTrendLine: !show})
                            }}>
                        {
                            (this.state.showTrendLine) ? 'Hide Trend' : 'Show Trend'
                        }      
                        </button>
                    </div>
                    <div className="chart-container">
                        <ResponsiveContainer height='100%' width='100%'>
                            <LineChart data={this.props.data} margin={{top: 5, right: 10, left: 30, bottom: 5}}>
                                <XAxis dataKey="timestamp"/>
                                <YAxis tickSize={10} dataKey="amount" domain={["dataMin", "dataMax"]} allowDataOverflow={true} />
                                <CartesianGrid strokeDasharray="3 3" stroke="#FFDDED"/>
                                <Tooltip/>
                                <Legend />
                                {
                                    (this.state.showTrendLine) && (
                                        <Line dot={false} type="monotone" dataKey="trend" stroke="#E477D4" strokeWidth={4}/>
                                    )
                                }
                                <Line dot={false} type="monotone" dataKey="amount" stroke="#7F4276"/>
                            </LineChart>
                        </ResponsiveContainer>
                    </div>
                </div>
            </div>
        );
    };
}

const convertToExchange = (exchangeRate, cryptos) => {
    const converted = cryptos.map((entry) => ({
        amount: (parseFloat(entry.amount) * exchangeRate).toFixed(2),
        timestamp: entry.timestamp
    }));

    return converted;
}

const getTrendData = (crypto) => {
    let x = [];
    
    for(let i = 0; i < crypto.length; i++) {
        x.push(i);
    }

    const xSum = x.map(num => num).reduce((prev, next) => prev + next);
    const ySum = crypto.map(entry => parseFloat(entry.amount)).reduce((prev, next) => prev + next);
    const averageX = xSum / crypto.length;
    const averageY = ySum / crypto.length;

    const topSlopeArray = [];
    for(let i = 0; i < crypto.length; i++) {
        topSlopeArray.push((x[i] - averageX) * (crypto[i].amount - averageY));
    }
    const topSlope = topSlopeArray.reduce((prev, next) => prev + next);
    const bottomSlope = x.map(num => (num - averageX) * (num - averageX)).reduce ((prev, next) => prev + next);

    const slope = topSlope / bottomSlope;
    const yIntercept = averageY - (slope * averageX);
    const percentage = (crypto[crypto.length - 1].amount - crypto[0].amount) / crypto[0].amount * 100;

    const data = crypto.map((entry, index) => ({
                timestamp: moment(entry.timestamp).format('ha'), 
                amount: parseFloat(entry.amount), 
                trend: slope * index + yIntercept
            }));

    return {
        data,
        percentage,
        current: crypto[crypto.length - 1]
    };
};

const mapStateToProps = (state, props) => {

    let crypto = selectCryptos(state.cryptos, state.filters)[props.name.toLowerCase()];

    const exchangeRate = parseFloat(state.filters.currency.exchange_rate);

    //converting amounts to exchange rate
    crypto = convertToExchange(exchangeRate, crypto);

    return {
        ...getTrendData(crypto),
        currency_symbol: state.filters.currency.symbol
    }
    
};

export default connect(mapStateToProps)(CryptoItem);