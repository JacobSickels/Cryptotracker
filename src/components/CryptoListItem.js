import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import momentTimezone from 'moment-timezone';


const CryptoListItem = (props) => (
    <div>
    {
        props.crypto.map(
            (entry) => <p key={entry.timestamp}> Amount: { entry.amount } Time: { moment(entry.timestamp).tz('America/Chicago').format() } </p>
        ) 
    }
    </div>
);
export default CryptoListItem;