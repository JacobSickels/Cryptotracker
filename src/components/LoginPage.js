import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="login">
        <div className="login__box">
            <div className="login__image">
                <img src="/images/loader.gif" />
            </div>
            <h1 className="login__title">CryptoTracker</h1>
            <p className="login__subtitle">Tracking cryptos, one hour at a time.</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
        </div>    
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);