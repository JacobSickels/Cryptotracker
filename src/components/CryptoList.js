import React from 'react';
import { connect } from 'react-redux';
import selectCryptos from '../selectors/cryptos';

import CryptoListItem from './CryptoListItem';

export const CryptoList = (props) => (
    <div> 
        <h2>CryptoList</h2> 
        Bitcoin:
        <CryptoListItem crypto={props.cryptos.bitcoin} />
        Litecoin:
        <CryptoListItem crypto={props.cryptos.litecoin} /> 
        Ethereum:
        <CryptoListItem crypto={props.cryptos.ethereum} />
    </div>
);

const mapStateToProps = (state) => {
    return {
        cryptos: selectCryptos(state.cryptos, state.filters)
    };
};

export default connect(mapStateToProps)(CryptoList);