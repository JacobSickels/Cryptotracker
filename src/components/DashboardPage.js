import React from 'react';
import { connect } from 'react-redux';
import CryptoList from './CryptoList';
import { setDefaultFilter } from '../actions/filters';
import { startSetCryptos } from '../actions/cryptos';

class DashboardPage extends React.Component {

    constructor(props) {
        super(props);
    }

    //Called before the component renders
    componentWillMount() {
        //Dashboard should only show the current day of data. 
        //This dispatch resets the filters to the current day
        this.props.setDefaultFilter();
    }

    render() {
        return (
            <div>
                <div className="page-container">
                    <CryptoList />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setDefaultFilter: () => dispatch(setDefaultFilter())
});

export default connect(null, mapDispatchToProps)(DashboardPage);