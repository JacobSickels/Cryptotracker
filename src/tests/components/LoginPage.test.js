import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

//This test makes sure the LoginPage component is rendered correctly
test('Should render LoginPage correctly', () => {
    const wrapper = shallow(<LoginPage startLogin={() => {}} />);
    expect(wrapper).toMatchSnapshot();
});

// This test makes sure the startLogin dispatch function is called 
// when the Login button is clicked
test('Use Case #2.0: Should call startLogin on button click', () => {
    const startLogin = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={startLogin} />);
    wrapper.find('button').simulate('click');
    expect(startLogin).toHaveBeenCalled();
});