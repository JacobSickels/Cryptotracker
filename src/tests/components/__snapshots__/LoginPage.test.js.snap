// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage correctly 1`] = `
<div
  className="login"
>
  <div
    className="login__box"
  >
    <div
      className="login__image"
    >
      <img
        src="/images/Crypto_Logo.svg"
      />
    </div>
    <h1
      className="login__title"
    >
      CryptoTracker
    </h1>
    <p
      className="login__subtitle"
    >
      Tracking cryptos, one hour at a time.
    </p>
    <button
      className="button"
      onClick={[Function]}
    >
      Login with Google
    </button>
  </div>
</div>
`;
